common/interpreter/error.go

```diff
@@ -0,0 +1,34 @@
+package interpreter
+
+import (
+	"errors"
+	"fmt"
+	"mime"
+	"net/http"
+
+	"github.com/amp-labs/connectors/common"
+)
+
+var ErrUnmarshal = errors.New("unmarshal failed")
+
+// FaultyResponseHandler used to parse erroneous response.
+type FaultyResponseHandler func(res *http.Response, body []byte) error
+
+// ErrorHandler invokes a function that matches response media type with parse error, ex: JSON<->JsonParserMethod
+// otherwise defaults to general error interpretation.
+type ErrorHandler struct {
+	JSON FaultyResponseHandler
+}
+
+func (h ErrorHandler) Handle(res *http.Response, body []byte) error {
+	mediaType, _, err := mime.ParseMediaType(res.Header.Get("Content-Type"))
+	if err != nil {
+		return fmt.Errorf("mime.ParseMediaType failed: %w", err)
+	}
+
+	if h.JSON != nil && mediaType == "application/json" {
+		return h.JSON(res, body)
```

Comment:
Depending on a ContentType a connector can parse error response in JSON, XML, HTML whatever impl we want to provide.
Every connector understands JSON structure in their own way. Other media types could be added.